File: app/src/main/java/pl/rpieja/flat/ExpensesTab.java
package pl.rpieja.flat;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.TextView;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;

import pl.rpieja.flat.dto.Income;
import pl.rpieja.flat.viewmodels.ChargesViewModel;

/**
 * Created by radix on 2017-11-01.
 */

public class ExpensesTab extends Fragment {
    public ListView listView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.expenses_tab, container, false);
        listView = rootView.findViewById(R.id.expensesListView);
        ChargesViewModel chargesViewModel = ViewModelProviders.of(getActivity()).get(ChargesViewModel.class);

        updateListWithCharges(chargesViewModel.getIncomesList());

        return rootView;
    }

    private static String CapitalizeFirstLetter(String input) {
        return input.substring(0, 1).toUpperCase() + input.substring(1);
    }

    private static Double round(Double value, int places) {
        if (places < 0) throw new IllegalArgumentException();

        BigDecimal bd = new BigDecimal(value);
        bd = bd.setScale(places, RoundingMode.HALF_UP);
        return bd.doubleValue();
    }

    public void updateListWithCharges(final List<Income> incomes) {
        listView.setAdapter(new BaseAdapter() {
            @Override
            public int getCount() {
                return incomes.size();
            }

            @Override
            public Object getItem(int i) {
                return incomes.get(i);
            }

            @Override
            public long getItemId(int i) {
                return 0;
            }

            @Override
            public View getView(int i, View view, ViewGroup viewGroup) {
                view = getLayoutInflater().inflate(R.layout.charges_item, null);

                TextView chargeName = view.findViewById(R.id.chargeName);
                TextView chargeAmount = view.findViewById(R.id.chargeAmount);
                TextView chargeUsers = view.findViewById(R.id.chargeUsers);

                String test = incomes.get(i).from.name;

                chargeName.setText(CapitalizeFirstLetter(incomes.get(i).name));
                chargeAmount.setText(round(incomes.get(i).amount, 2).toString());
                chargeUsers.setText(test);

                return view;
            }
        });
    }
}

File: app/src/main/java/pl/rpieja/flat/ExpensesTab.kt
package pl.rpieja.flat

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import pl.rpieja.flat.dto.Expense
import pl.rpieja.flat.viewmodels.ChargesViewModel
import java.text.NumberFormat
import java.util.*

class ExpensesTab : Fragment() {
    private class ItemViewHolder(itemView: View): RecyclerView.ViewHolder(itemView) {
        val chargeName: TextView = itemView.findViewById(R.id.chargeName)
        val chargeAmount: TextView = itemView.findViewById(R.id.chargeAmount)
        val chargeUsers: TextView = itemView.findViewById(R.id.chargeUsers)
    }

    private class ItemAdapter(private val expenses: List<Expense>):
            RecyclerView.Adapter<ItemViewHolder>() {
        private val currencyFormat = NumberFormat.getCurrencyInstance()

        init {
            currencyFormat.currency = Currency.getInstance("USD") // TODO fetch
        }

        override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ItemViewHolder {
            val inflater = LayoutInflater.from(parent!!.context)
            return ItemViewHolder(inflater.inflate(R.layout.charges_item, parent, false))
        }

        override fun onBindViewHolder(holder: ItemViewHolder?, position: Int) {
            val income = expenses[position]
            holder!!.chargeName.text = income.name
            holder.chargeAmount.text = currencyFormat.format(income.amount).toString()
            holder.chargeUsers.text = income.from.name
        }

        override fun getItemCount(): Int = expenses.size
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.expenses_tab, container, false)
        val recyclerView: RecyclerView = rootView.findViewById(R.id.expensesListView)
        val chargesViewModel = ViewModelProviders.of(activity!!).get(ChargesViewModel::class.java)

        recyclerView.layoutManager = LinearLayoutManager(activity)

        chargesViewModel.chargesList.observe(this, Observer { chargesDTO ->
            recyclerView.adapter = ItemAdapter(chargesDTO!!.incomes)
        })

        return rootView
    }
}
